boards native_sim: Deprecate CONFIG_NATIVE_APPLICATION & CONFIG_NATIVE_SIM_NATIVE_POSIX_COMPAT

aescolar opened the pull request:
boards native: Deprecate CONFIG_NATIVE_APPLICATION
    
    This option is used in tree only by native_posix, which is deprecated
    and being replaced by native_sim. But may be used also in out of tree
    targets.
    As part of the native_posix deprecation, and therefore the lack of
    testing this feature would have in the future, we are also deprecating
    this option.

----

    boards native_sim: Deprecate CONFIG_NATIVE_SIM_NATIVE_POSIX_COMPAT
    
    This option existed only to make the transition from native_posix to
    native_sim easier. As native_posix is going to be removed in v4.2
    we deprecate this option now, so it will also be removed.
    
    We also switch this option to default to false already now.

& add to release notes and migration guide

aescolar committed:
boards native: Deprecate CONFIG_NATIVE_APPLICATION

This option is used in tree only by native_posix, which is deprecated
and being replaced by native_sim. But may be used also in out of tree
targets.
As part of the native_posix deprecation, and therefore the lack of
testing this feature would have in the future, we are also deprecating
this option.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
boards native_sim: Deprecate CONFIG_NATIVE_SIM_NATIVE_POSIX_COMPAT

This option existed only to make the transition from native_posix to
native_sim easier. As native_posix is going to be removed in v4.2
we deprecate this option now, so it will also be removed.

We also switch this option to default to false already now.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
doc: 4.1 release & migration guide: Add native deprecation

Mention that CONFIG_NATIVE_APPLICATION &
CONFIG_NATIVE_SIM_NATIVE_POSIX_COMPAT have been deprecated.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

kartben commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -110,3 +110,11 @@ Modem
+    use this option should migrate to use the native_simulator as their runner (:github:`81232`).
+  * For the native_sim target :kconfig:option:`CONFIG_NATIVE_SIM_NATIVE_POSIX_COMPAT` has been
+    switched to ``n`` by default, and this option has been deprecated. Ensure your code does not
+    use anymore on the :kconfig:option:`CONFIG_BOARD_NATIVE_POSIX` option (:github:`81232`).
```
```suggestion
    use the :kconfig:option:`CONFIG_BOARD_NATIVE_POSIX` option anymore (:github:`81232`).
```

kartben commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -110,3 +110,11 @@ Modem
+
+* native/POSIX
+
+  * :kconfig:option:`CONFIG_NATIVE_APPLICATION` has been deprecated. Out of tree boards which
+    use this option should migrate to use the native_simulator as their runner (:github:`81232`).
```
```suggestion
  * :kconfig:option:`CONFIG_NATIVE_APPLICATION` has been deprecated. Out-of-tree boards using
    this option should migrate to the native_simulator runner (:github:81232).
```

rugeGerritsen commented on boards/native/native_sim/Kconfig:
```diff
@@ -18,12 +18,14 @@ comment "Native Simular (Single Core) options"
 config NATIVE_SIM_NATIVE_POSIX_COMPAT
 	bool "Pretend to be a native_posix board"
-	default y
+	default n
```
The Kconfig guidelines state that we should not use `default n`. I'm fine with leaving it as is

aescolar replied:
removed default :)

aescolar commented:
3 new pushes since last:
1. removed default kconfig as requested by @rugeGerritsen 
2. rebase on top of main without changes
3. added a link in the migration guide with an example of how to convert a board from being based on NATIVE_APPLICATION to NATIVE_LIBRARY
